### PR TITLE
Merging Boundary Conditions  and Test Cases to main

### PR DESCRIPTION
**Pull request by hannahwhitby**

### CODE REVIEW

File: Test_Cases.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
+#import functions for spatial grid, time grid, upstream step, boundary conditions, loaf an interpolate IC
+from src.solver_functions import()
```
Comment (sagebidwell) [inline Test_Cases.py]: Please could you check import() as I am unsure if this is valid for python, I think that it should be import *

File: Test_Cases.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+plt.legend()
+plt.suptitle("Test Case 3: Sensitivity Analysis")
+plt.tight_layout()
+plt.show()
+
+#test case 4
+tau = 100
+decay = np.exp(-t / tau)
+theta0[0] *= decay[0]
```
Comment (sagebidwell) [inline Test_Cases.py]: Do you definitely need this line?

File: Test_Cases.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+from src.solver_functions import()
+
+#using plot_snapshots to create consistent plots without repeating lots of code
+def plot_snapshots(results, x, dt, title):
+    nt = results.shape[0]
+    times = [0, nt//3, 2*nt//3, nt-1] #choosing times at 1/3 intervals
+
+    plt.figure(figsize=(8,4))
+    for ti in times:
+        plt.plot(x, results[ti], label=f"t"{ti*dt:.0f}s")
```
[the remark this answers is not in the document]
Comment (hannahwhitby) [reply Test_Cases.py]: ok thank you

File: Test_Cases.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    plt.title(title)
+    plt.xlabel("Distance (m)")
+    plt.ylabel("Concentration (µg/m^3)")
+    plt.grid(True)   
+    plt.legend()
+    plt.tight_layout()
+    plt.show()
+
+#to avoid repeating time loops
+def run_simulation(theta0, u, x, t, BC="sero_gradient"):
```
[the remark this answers is not in the document]
Comment (hannahwhitby) [reply Test_Cases.py]: Thank you!!!!!

File: Test_Cases.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    dt = t[1] - t[0]
+    nt = len(t)
+
+    results = np.zeros((nt, len(x)))
+    results[0] = theta0.copy()
+    theta = theta0.copy()
+    
+    for n in range(1, nt):
+        theta, _ = upstream_step(theta, u, dx, dt)
+        theta = apply_boundary_conditions(theta, BC)
```
[the remark this answers is not in the document]
Comment (hannahwhitby) [reply Test_Cases.py]: ok perfect, that makes sense based off the boundary conditions script

File: Test_Cases.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+#for test cases 1,4,5
+theta0 = np.zeros_like(x)
+theta0[0] = 250
+
+#test case one
+x, g = model_spatial_grid(20, 0.2)
+t, ti = model_time_grid(300, 10)
+
+C1 = run_simulation(theta0, u=0.1, x=x, t=t)
+plot_snapshots(C1, x, dt=10, title="Test Case 1: Delta Initial Condition"))
```
[the remark this answers is not in the document]
Comment (hannahwhitby) [reply Test_Cases.py]: Will get that all changed now!

File: Test_Cases.py (hunk 2 above)
Comment (hannahwhitby) [reply Test_Cases.py]: Because this test involves exponential decay starting from t=0 I think I should keep it

File: Test_Cases.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+C4 = run_simulation(theta0, u=0.1, x=x, t=t)
+C4[:, 0] *= decay
+plot_snapshots(C4, x, dt=10, title="Test Case 4: Exponential Decay")
+
+#test case 5
+rng = np.random.default_rng(123)
+u_pert = 0.1 * (1 + 0.1 * rng.standard.normal(len(x)))
+
+C5 = run_simulation(theta0, u_pert, x=x, t=t)
+plot_snapshots(C5, x, dt=10, title="Test Case 5: Perturbed Verlocity")
```
[the remark this answers is not in the document]
Comment (hannahwhitby) [reply Test_Cases.py]: Thank you.
